In [2]:
import sys

print(sys.executable)
print(sys.version)

c:\Users\Admin\OneDrive\Desktop\SWYNEX-Data-Cleaning-Preparation\.venv-1\Scripts\python.exe
3.12.4 (tags/v3.12.4:8e8a4ba, Jun  6 2024, 19:30:16) [MSC v.1940 64 bit (AMD64)]


In [3]:
import sys
!{sys.executable} -m pip install pandas numpy matplotlib seaborn openpyxl jupyter


[notice] A new release of pip is available: 24.0 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [4]:
import pandas as pd

print(pd.__version__)

3.0.6


In [5]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("All libraries imported successfully!")

All libraries imported successfully!


In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

print("Libraries imported successfully!")

Libraries imported successfully!


In [7]:
file_path = "../data/raw/online_retail_II.xlsx"

# Check the available sheets
excel_file = pd.ExcelFile(file_path)

print("Available sheets:")
print(excel_file.sheet_names)

Available sheets:
['Year 2009-2010', 'Year 2010-2011']


In [8]:
df_2009_2010 = pd.read_excel(
    file_path,
    sheet_name="Year 2009-2010"
)

df_2010_2011 = pd.read_excel(
    file_path,
    sheet_name="Year 2010-2011"
)

In [9]:
print("2009-2010:", df_2009_2010.shape)
print("2010-2011:", df_2010_2011.shape)

2009-2010: (525461, 8)
2010-2011: (541910, 8)


In [10]:
df = pd.concat(
    [df_2009_2010, df_2010_2011],
    ignore_index=True
)

print("Combined dataset:", df.shape)

Combined dataset: (1067371, 8)


In [11]:
df.head()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [12]:
df.tail()

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
1067366,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
1067367,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
1067368,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
1067369,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.0,France
1067370,581587,POST,POSTAGE,1,2011-12-09 12:50:00,18.00,12680.0,France


In [13]:
print("Rows:", df.shape[0])
print("Columns:", df.shape[1])

Rows: 1067371
Columns: 8


In [14]:
print(df.dtypes)

Invoice                object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
Price                 float64
Customer ID           float64
Country                   str
dtype: object


In [15]:
missing_values = df.isnull().sum()

print(missing_values)

Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64


In [16]:
missing_percentage = (df.isnull().sum() / len(df)) * 100

missing_summary = pd.DataFrame({
    "Missing Values": missing_values,
    "Missing Percentage": missing_percentage
})

missing_summary = missing_summary.sort_values(
    by="Missing Values",
    ascending=False
)

missing_summary

,Missing Values,Missing Percentage
Customer ID,243007,22.766873
Description,4382,0.410541
StockCode,0,0.000000
Invoice,0,0.000000
Quantity,0,0.000000
InvoiceDate,0,0.000000
Price,0,0.000000
Country,0,0.000000


In [17]:
duplicate_count = df.duplicated().sum()

print("Number of duplicate records:", duplicate_count)

Number of duplicate records: 34335


In [18]:
duplicate_percentage = (
    duplicate_count / len(df)
) * 100

print(
    f"Duplicate percentage: {duplicate_percentage:.2f}%"
)

Duplicate percentage: 3.22%


In [19]:
duplicates = df[df.duplicated(keep=False)]

duplicates.head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
367,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom


In [20]:
print("Quantity statistics:")
print(df["Quantity"].describe())

print("\nPrice statistics:")
print(df["Price"].describe())

Quantity statistics:
count    1.067371e+06
mean     9.938898e+00
std      1.727058e+02
min     -8.099500e+04
25%      1.000000e+00
50%      3.000000e+00
75%      1.000000e+01
max      8.099500e+04
Name: Quantity, dtype: float64

Price statistics:
count    1.067371e+06
mean     4.649388e+00
std      1.235531e+02
min     -5.359436e+04
25%      1.250000e+00
50%      2.100000e+00
75%      4.150000e+00
max      3.897000e+04
Name: Price, dtype: float64


In [21]:
negative_quantity = (df["Quantity"] < 0).sum()

print("Negative quantity records:", negative_quantity)

Negative quantity records: 22950


In [22]:
zero_quantity = (df["Quantity"] == 0).sum()

print("Zero quantity records:", zero_quantity)

Zero quantity records: 0


In [23]:
zero_price = (df["Price"] == 0).sum()

print("Zero price records:", zero_price)

Zero price records: 6202


In [24]:
negative_price = (df["Price"] < 0).sum()

print("Negative price records:", negative_price)

Negative price records: 5


In [25]:
cancelled_invoices = df["Invoice"].astype(str).str.startswith("C")

print(
    "Cancellation records:",
    cancelled_invoices.sum()
)

Cancellation records: 19494


In [26]:
df[cancelled_invoices].head(10)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321.0,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592.0,United Kingdom


In [27]:
print("Number of countries:", df["Country"].nunique())

print(df["Country"].unique())

Number of countries: 43
<StringArray>
[      'United Kingdom',               'France',                  'USA',
              'Belgium',            'Australia',                 'EIRE',
              'Germany',             'Portugal',                'Japan',
              'Denmark',              'Nigeria',          'Netherlands',
               'Poland',                'Spain',      'Channel Islands',
                'Italy',               'Cyprus',               'Greece',
               'Norway',              'Austria',               'Sweden',
 'United Arab Emirates',              'Finland',          'Switzerland',
          'Unspecified',                'Malta',              'Bahrain',
                  'RSA',              'Bermuda',            'Hong Kong',
            'Singapore',             'Thailand',               'Israel',
            'Lithuania',          'West Indies',              'Lebanon',
                'Korea',               'Brazil',               'Canada',
             

In [28]:
print("Number of unique stock codes:", df["StockCode"].nunique())
print("Number of unique invoices:", df["Invoice"].nunique())
print("Number of unique customers:", df["Customer ID"].nunique())

Number of unique stock codes: 5305
Number of unique invoices: 53628
Number of unique customers: 5942


In [29]:
data_quality_report = pd.DataFrame({
    "Column": df.columns,
    "Data Type": df.dtypes.astype(str).values,
    "Missing Values": df.isnull().sum().values,
    "Missing Percentage": (
        df.isnull().sum().values / len(df) * 100
    ).round(2),
    "Unique Values": [
        df[column].nunique(dropna=True)
        for column in df.columns
    ]
})

data_quality_report

,Column,Data Type,Missing Values,Missing Percentage,Unique Values
0,Invoice,object,0,0.00,53628
1,StockCode,object,0,0.00,5305
2,Description,object,4382,0.41,5698
3,Quantity,int64,0,0.00,1057
4,InvoiceDate,datetime64[us],0,0.00,47635
5,Price,float64,0,0.00,2807
6,Customer ID,float64,243007,22.77,5942
7,Country,str,0,0.00,43


## Initial Data Quality Findings

The raw Online Retail II dataset contains 1,067,371 transaction records and 8 columns.

Initial inspection identified the following data-quality issues:

1. Customer ID contains 243,007 missing values (22.77%).
2. Description contains 4,382 missing values (0.41%).
3. The dataset contains 34,335 exact duplicate records.
4. There are 22,950 records with negative quantities.
5. There are 6,202 records with a price of zero.
6. There are 5 records with negative prices.
7. Customer ID is stored as a floating-point data type because of missing values and requires appropriate type handling.
8. InvoiceDate is already stored as a datetime data type.

In [30]:
negative_qty_df = df[df["Quantity"] < 0]

negative_qty_df.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321.0,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321.0,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592.0,United Kingdom


In [31]:
print(
    "Negative quantity records:",
    len(negative_qty_df)
)

print(
    "Negative quantity with cancelled invoice:",
    negative_qty_df["Invoice"].astype(str).str.startswith("C").sum()
)

Negative quantity records: 22950
Negative quantity with cancelled invoice: 19493


In [32]:
print(
    "Negative quantity without cancelled invoice:",
    (
        ~negative_qty_df["Invoice"].astype(str).str.startswith("C")
    ).sum()
)

Negative quantity without cancelled invoice: 3457


In [33]:
zero_price_df = df[df["Price"] == 0]
zero_price_df.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,NaN,United Kingdom
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,NaN,United Kingdom
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom


In [34]:
print("Zero-price records:", len(zero_price_df))

print("\nDescription:")
print(zero_price_df["Description"].value_counts().head(20)) 

Zero-price records: 6202

Description:
Description
check                                  162
?                                       92
damages                                 84
damaged                                 81
found                                   28
missing                                 27
sold as set on dotcom                   20
Damaged                                 17
adjustment                              16
OWL DOORSTOP                            15
POLYESTER FILLER PAD 45x45cm            12
dotcom                                  12
amazon                                  11
POLYESTER FILLER PAD 40x40cm            10
IVORY KITCHEN SCALES                    10
FRENCH BLUE METAL DOOR SIGN 1           10
smashed                                  9
Found                                    9
PICNIC BASKET WICKER LARGE               9
AIRLINE BAG VINTAGE WORLD CHAMPION       9
Name: count, dtype: int64


In [35]:
print("\nStock codes:")
print(zero_price_df["StockCode"].value_counts().head(20))


Stock codes:
StockCode
46000M    18
22501     18
79321     17
21116     16
22423     16
23084     16
46000S    15
22734     15
22139     14
35965     14
84990     13
22502     13
20713     13
22469     12
84016     11
22627     11
71477     10
22470     10
22625     10
22624     10
Name: count, dtype: int64


In [36]:
negative_price_df = df[df["Price"] < 0]
negative_price_df

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom
825444,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
825445,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [37]:
negative_price_df.T

,179403,276274,403472,825444,825445
Invoice,A506401,A516228,A528059,A563186,A563187
StockCode,B,B,B,B,B
Description,Adjust bad debt,Adjust bad debt,Adjust bad debt,Adjust bad debt,Adjust bad debt
Quantity,1,1,1,1,1
InvoiceDate,2010-04-29 13:36:00,2010-07-19 11:24:00,2010-10-20 12:04:00,2011-08-12 14:51:00,2011-08-12 14:52:00
Price,-53594.36,-44031.79,-38925.87,-11062.06,-11062.06
Customer ID,NaN,NaN,NaN,NaN,NaN
Country,United Kingdom,United Kingdom,United Kingdom,United Kingdom,United Kingdom


In [38]:
print("Negative price records:")
print(negative_price_df)

Negative price records:
        Invoice StockCode      Description  Quantity         InvoiceDate  \
179403  A506401         B  Adjust bad debt         1 2010-04-29 13:36:00   
276274  A516228         B  Adjust bad debt         1 2010-07-19 11:24:00   
403472  A528059         B  Adjust bad debt         1 2010-10-20 12:04:00   
825444  A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
825445  A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

           Price  Customer ID         Country  
179403 -53594.36          NaN  United Kingdom  
276274 -44031.79          NaN  United Kingdom  
403472 -38925.87          NaN  United Kingdom  
825444 -11062.06          NaN  United Kingdom  
825445 -11062.06          NaN  United Kingdom  


In [39]:
duplicates = df[df.duplicated(keep=False)]
duplicates.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
367,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom


In [40]:
duplicates.sort_values(
    by=["Invoice", "StockCode"]
).head(30)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
391,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
386,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
394,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom


In [41]:
print(
    "Duplicate rows:",
    df.duplicated().sum()
)

print(
    "Percentage duplicated:",
    round(df.duplicated().mean() * 100, 2),
    "%"
)

Duplicate rows: 34335
Percentage duplicated: 3.22 %


In [42]:
missing_customer = df[df["Customer ID"].isna()]
missing_customer.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.00,NaN,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.00,NaN,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.00,NaN,United Kingdom
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.00,NaN,United Kingdom
577,489525,85226C,BLUE PULL BACK RACING CAR,1,2009-12-01 11:49:00,0.55,NaN,United Kingdom
578,489525,85227,SET/6 3D KIT CARDS FOR KIDS,1,2009-12-01 11:49:00,0.85,NaN,United Kingdom
1055,489548,22271,FELTCRAFT DOLL ROSIE,1,2009-12-01 12:32:00,2.95,NaN,United Kingdom
1056,489548,22254,FELT TOADSTOOL LARGE,12,2009-12-01 12:32:00,1.25,NaN,United Kingdom
1057,489548,22273,FELTCRAFT DOLL MOLLY,3,2009-12-01 12:32:00,2.95,NaN,United Kingdom
1058,489548,22195,LARGE HEART MEASURING SPOONS,1,2009-12-01 12:32:00,1.65,NaN,United Kingdom


In [43]:
print(
    "Rows with missing Customer ID:",
    len(missing_customer)
)

print(
    "Percentage:",
    round(
        len(missing_customer) / len(df) * 100,
        2
    ),
    "%"
)

Rows with missing Customer ID: 243007
Percentage: 22.77 %


In [44]:
print(
    "Missing Customer ID + cancelled invoice:",
    (
        missing_customer["Invoice"]
        .astype(str)
        .str.startswith("C")
        .sum()
    )
)

Missing Customer ID + cancelled invoice: 750


In [45]:
missing_description = df[df["Description"].isna()]
missing_description.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
3161,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom
3731,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom
4566,489821,85049G,NaN,-240,2009-12-02 13:25:00,0.0,NaN,United Kingdom
6378,489882,35751C,NaN,12,2009-12-02 16:22:00,0.0,NaN,United Kingdom
6555,489898,79323G,NaN,954,2009-12-03 09:40:00,0.0,NaN,United Kingdom
6576,489901,21098,NaN,-200,2009-12-03 09:47:00,0.0,NaN,United Kingdom
6581,489903,21166,NaN,48,2009-12-03 09:57:00,0.0,NaN,United Kingdom


In [46]:
missing_description[
    ["Invoice", "StockCode", "Description", "Quantity", "Price", "Customer ID"]
].head(20)

,Invoice,StockCode,Description,Quantity,Price,Customer ID
470,489521,21646,NaN,-50,0.0,NaN
3114,489655,20683,NaN,-44,0.0,NaN
3161,489659,21350,NaN,230,0.0,NaN
3731,489781,84292,NaN,17,0.0,NaN
4296,489806,18010,NaN,-770,0.0,NaN
4566,489821,85049G,NaN,-240,0.0,NaN
6378,489882,35751C,NaN,12,0.0,NaN
6555,489898,79323G,NaN,954,0.0,NaN
6576,489901,21098,NaN,-200,0.0,NaN
6581,489903,21166,NaN,48,0.0,NaN


In [47]:
test_columns = ["Invoice", "StockCode", "Description", "Country"]

for column in test_columns:
    print(f"\n--- {column} ---")
    print("Leading and trailing whitespace:",
          df[column].astype(str).str.contains(r"^\s|\s$", regex=True).sum())


--- Invoice ---
Leading and trailing whitespace: 0

--- StockCode ---
Leading and trailing whitespace: 1

--- Description ---
Leading and trailing whitespace: 213035

--- Country ---
Leading and trailing whitespace: 0


In [48]:
print("Number of countries:", df["Country"].nunique())

print(df["Country"].value_counts())

Number of countries: 43
Country
United Kingdom          981330
EIRE                     17866
Germany                  17624
France                   14330
Netherlands               5140
Spain                     3811
Switzerland               3189
Belgium                   3123
Portugal                  2620
Australia                 1913
Channel Islands           1664
Italy                     1534
Norway                    1455
Sweden                    1364
Cyprus                    1176
Finland                   1049
Austria                    938
Denmark                    817
Unspecified                756
Greece                     663
Japan                      582
USA                        535
Poland                     535
United Arab Emirates       500
Israel                     371
Hong Kong                  364
Singapore                  346
Malta                      299
Iceland                    253
Canada                     228
Lithuania                  189
RSA    

In [49]:
print(
    "Unique descriptions:",
    df["Description"].nunique()
)

Unique descriptions: 5698


In [50]:
df["Customer ID"] = df["Customer ID"].astype("Int64")

In [51]:
cancelled_mask = df["Invoice"].astype(str).str.startswith("C")

print(
    "Cancelled invoices/records:",
    cancelled_mask.sum()
)

print(
    "Percentage:",
    round(cancelled_mask.mean() * 100, 2),
    "%"
)

Cancelled invoices/records: 19494
Percentage: 1.83 %


In [52]:
cancelled_df = df[cancelled_mask]
cancelled_df.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
178,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321,Australia
179,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321,Australia
180,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321,Australia
181,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321,Australia
182,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321,Australia
183,C489449,21871,SAVE THE PLANET MUG,-12,2009-12-01 10:33:00,1.25,16321,Australia
184,C489449,84946,ANTIQUE SILVER TEA GLASS ETCHED,-12,2009-12-01 10:33:00,1.25,16321,Australia
185,C489449,84970S,HANGING HEART ZINC T-LIGHT HOLDER,-24,2009-12-01 10:33:00,0.85,16321,Australia
186,C489449,22090,PAPER BUNTING RETRO SPOTS,-12,2009-12-01 10:33:00,2.95,16321,Australia
196,C489459,90200A,PURPLE SWEETHEART BRACELET,-3,2009-12-01 10:44:00,4.25,17592,United Kingdom


In [53]:
print(
    cancelled_df["Quantity"].describe()
)

count    19494.000000
mean       -25.186827
std        805.104908
min     -80995.000000
25%         -6.000000
50%         -2.000000
75%         -1.000000
max          1.000000
Name: Quantity, dtype: float64


In [54]:
print(
    "Negative quantity records:",
    len(negative_qty_df)
)

print(
    "Negative quantity with cancelled invoice:",
    negative_qty_df["Invoice"].astype(str).str.startswith("C").sum()
)

print(
    "Negative quantity without cancelled invoice:",
    (~negative_qty_df["Invoice"].astype(str).str.startswith("C")).sum()
)

Negative quantity records: 22950
Negative quantity with cancelled invoice: 19493
Negative quantity without cancelled invoice: 3457


In [55]:
print("Zero-price records:", len(zero_price_df))
print(zero_price_df["Description"].value_counts().head(10))

Zero-price records: 6202
Description
check                    162
?                         92
damages                   84
damaged                   81
found                     28
missing                   27
sold as set on dotcom     20
Damaged                   17
adjustment                16
OWL DOORSTOP              15
Name: count, dtype: int64


In [56]:
negative_price_df

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
179403,A506401,B,Adjust bad debt,1,2010-04-29 13:36:00,-53594.36,NaN,United Kingdom
276274,A516228,B,Adjust bad debt,1,2010-07-19 11:24:00,-44031.79,NaN,United Kingdom
403472,A528059,B,Adjust bad debt,1,2010-10-20 12:04:00,-38925.87,NaN,United Kingdom
825444,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
825445,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [57]:
duplicates.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
362,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
363,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
365,489517,21821,GLITTER STAR GARLAND WITH BELLS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
367,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
368,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
371,489517,21912,VINTAGE SNAKES & LADDERS,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom
379,489517,21491,SET OF THREE VINTAGE GIFT WRAPS,1,2009-12-01 11:34:00,1.95,16329.0,United Kingdom
383,489517,22130,PARTY CONE CHRISTMAS DECORATION,6,2009-12-01 11:34:00,0.85,16329.0,United Kingdom
384,489517,22319,HAIRCLIPS FORTIES FABRIC ASSORTED,12,2009-12-01 11:34:00,0.65,16329.0,United Kingdom
385,489517,21913,VINTAGE SEASIDE JIGSAW PUZZLES,1,2009-12-01 11:34:00,3.75,16329.0,United Kingdom


In [58]:
print("Cancelled records:", cancelled_mask.sum())
print("Cancelled + negative quantity:",
      (cancelled_mask & (df["Quantity"] < 0)).sum())

Cancelled records: 19494
Cancelled + negative quantity: 19493


In [59]:
missing_description = df[df["Description"].isna()]

print("Missing descriptions:", len(missing_description))

print(
    "Unique stock codes:",
    missing_description["StockCode"].nunique()
)

missing_description[
    ["StockCode", "Quantity", "Price", "Country"]
].head(20)

Missing descriptions: 4382
Unique stock codes: 2451


,StockCode,Quantity,Price,Country
470,21646,-50,0.0,United Kingdom
3114,20683,-44,0.0,United Kingdom
3161,21350,230,0.0,United Kingdom
3731,84292,17,0.0,United Kingdom
4296,18010,-770,0.0,United Kingdom
4566,85049G,-240,0.0,United Kingdom
6378,35751C,12,0.0,United Kingdom
6555,79323G,954,0.0,United Kingdom
6576,21098,-200,0.0,United Kingdom
6581,21166,48,0.0,United Kingdom


In [60]:
description_lookup = (
    df.dropna(subset=["Description"])
      .drop_duplicates("StockCode")
      .set_index("StockCode")["Description"]
)

missing_description_check = missing_description[
    ["StockCode", "Description"]
].copy()

missing_description_check["Known Description"] = (
    missing_description_check["StockCode"]
    .map(description_lookup)
)

missing_description_check.head(20)

,StockCode,Description,Known Description
470,21646,NaN,NaN
3114,20683,NaN,RAIN GIRL CHILDS UMBRELLA
3161,21350,NaN,GINGHAM HEART WREATH
3731,84292,NaN,ROSE PINK METAL FOLDING CHAIR
4296,18010,NaN,ASSORTED FRAGRANCE BATH CONFETTI
4566,85049G,NaN,CHOCOLATE BOX RIBBONS
6378,35751C,NaN,PURPLE CURRENT CANDLE RING
6555,79323G,NaN,GOLD CHERRY LIGHTS
6576,21098,NaN,CHRISTMAS TOILET ROLL
6581,21166,NaN,COOK WITH WINE METAL SIGN


In [61]:
negative_non_cancelled = df[
    (df["Quantity"] < 0) &
    (~df["Invoice"].astype(str).str.startswith("C"))
]

negative_non_cancelled.head(20)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
263,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,<NA>,United Kingdom
283,489463,71477,short,-240,2009-12-01 10:52:00,0.0,<NA>,United Kingdom
284,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,<NA>,United Kingdom
470,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,<NA>,United Kingdom
3114,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,<NA>,United Kingdom
3162,489660,35956,lost,-1043,2009-12-01 17:43:00,0.0,<NA>,United Kingdom
3168,489663,35605A,damages,-117,2009-12-01 18:02:00,0.0,<NA>,United Kingdom
4296,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,<NA>,United Kingdom
4538,489820,21133,invcd as 84879?,-720,2009-12-02 13:23:00,0.0,<NA>,United Kingdom
4566,489821,85049G,NaN,-240,2009-12-02 13:25:00,0.0,<NA>,United Kingdom


In [62]:
print(
    "Negative quantity without C invoice:",
    len(negative_non_cancelled)
)

print(
    negative_non_cancelled["Description"]
    .value_counts()
    .head(20)
)

Negative quantity without C invoice: 3457
Description
check                     123
damages                    84
?                          83
damaged                    78
missing                    27
sold as set on dotcom      20
Damaged                    17
smashed                     9
thrown away                 9
Unsaleable, destroyed.      9
dotcom                      8
damages?                    7
??                          7
crushed                     6
given away                  6
MIA                         5
Damages                     5
counted                     5
checked                     5
wet damaged                 5
Name: count, dtype: int64


## Data Cleaning Decisions

### 1. Duplicate Records
34,335 exact duplicate records were identified and will be removed because they contain identical information and do not provide additional transaction-level information.

### 2. Missing Customer IDs
243,007 records have missing Customer IDs. These records will be retained because other transaction attributes such as product, quantity, date, price, and country remain useful for analysis. Missing Customer IDs will remain as missing values.

### 3. Missing Product Descriptions
4,382 records have missing descriptions. The StockCode will be used to determine whether the description can be recovered from other transactions before deciding whether any records need to be removed.

### 4. Negative Quantities
22,950 records have negative quantities. These will not be automatically removed because negative quantities are associated with cancellations/returns in the dataset. A cancellation flag will be created for analysis.

### 5. Zero Prices
6,202 records have a price of zero. These records will be retained because some represent adjustments, damaged/missing items, or other non-standard transactions. They will be flagged for analysis.

### 6. Negative Prices
Five records contain negative prices and are associated with "Adjust bad debt" entries rather than normal product sales. These records will be excluded from the cleaned sales dataset.

### 7. Customer ID Data Type
Customer ID will be converted from float64 to Pandas nullable Int64 to represent customer identifiers correctly while preserving missing values.

In [63]:
missing_description = df[df["Description"].isna()]

print("Missing descriptions:", len(missing_description))

print(
    "Unique stock codes:",
    missing_description["StockCode"].nunique()
)

missing_description[
    ["StockCode", "Quantity", "Price", "Country"]
].head(20)

Missing descriptions: 4382
Unique stock codes: 2451


,StockCode,Quantity,Price,Country
470,21646,-50,0.0,United Kingdom
3114,20683,-44,0.0,United Kingdom
3161,21350,230,0.0,United Kingdom
3731,84292,17,0.0,United Kingdom
4296,18010,-770,0.0,United Kingdom
4566,85049G,-240,0.0,United Kingdom
6378,35751C,12,0.0,United Kingdom
6555,79323G,954,0.0,United Kingdom
6576,21098,-200,0.0,United Kingdom
6581,21166,48,0.0,United Kingdom


In [64]:
negative_non_cancelled = df[
    (df["Quantity"] < 0) &
    (~df["Invoice"].astype(str).str.startswith("C"))
]

print(
    "Negative quantity without C invoice:",
    len(negative_non_cancelled)
)

print(
    negative_non_cancelled["Description"]
    .value_counts()
    .head(20)
)

Negative quantity without C invoice: 3457
Description
check                     123
damages                    84
?                          83
damaged                    78
missing                    27
sold as set on dotcom      20
Damaged                    17
smashed                     9
thrown away                 9
Unsaleable, destroyed.      9
dotcom                      8
damages?                    7
??                          7
crushed                     6
given away                  6
MIA                         5
Damages                     5
counted                     5
checked                     5
wet damaged                 5
Name: count, dtype: int64


In [65]:
description_lookup = (
    df.dropna(subset=["Description"])
      .drop_duplicates("StockCode")
      .set_index("StockCode")["Description"]
)

missing_description_check = df[df["Description"].isna()][
    ["StockCode", "Description"]
].copy()

missing_description_check["Known Description"] = (
    missing_description_check["StockCode"]
    .map(description_lookup)
)

print(missing_description_check.head(20))

      StockCode Description                    Known Description
470       21646         NaN                                  NaN
3114      20683         NaN            RAIN GIRL CHILDS UMBRELLA
3161      21350         NaN                 GINGHAM HEART WREATH
3731      84292         NaN        ROSE PINK METAL FOLDING CHAIR
4296      18010         NaN     ASSORTED FRAGRANCE BATH CONFETTI
4566     85049G         NaN               CHOCOLATE BOX RIBBONS 
6378     35751C         NaN           PURPLE CURRENT CANDLE RING
6555     79323G         NaN                  GOLD  CHERRY LIGHTS
6576      21098         NaN                CHRISTMAS TOILET ROLL
6581      21166         NaN           COOK WITH WINE METAL SIGN 
7204      21982         NaN             PACK OF 12 SUKI TISSUES 
7205      21982         NaN             PACK OF 12 SUKI TISSUES 
7559      20620         NaN            RED SPOTTY PASSPORT COVER
8553      85064         NaN         CREAM SWEETHEART LETTER RACK
9249     84508B         N

In [66]:
recoverable_descriptions = (
    missing_description_check["Known Description"].notna().sum()
)

print(
    "Descriptions recoverable from StockCode:",
    recoverable_descriptions
)

print(
    "Descriptions still missing:",
    len(missing_description_check) - recoverable_descriptions
)

Descriptions recoverable from StockCode: 4019
Descriptions still missing: 363


In [67]:
df_raw = df.copy()

print("Raw rows:", len(df_raw))
print("Raw columns:", len(df_raw.columns))

Raw rows: 1067371
Raw columns: 8


In [68]:
df_clean = df_raw.drop_duplicates().copy()

print("Rows after removing duplicates:", len(df_clean))
print("Rows removed:", len(df_raw) - len(df_clean))

Rows after removing duplicates: 1033036
Rows removed: 34335


In [69]:
description_lookup = (
    df_clean.dropna(subset=["Description"])
            .drop_duplicates("StockCode")
            .set_index("StockCode")["Description"]
)

In [70]:
missing_before = df_clean["Description"].isna().sum()

df_clean["Description"] = (
    df_clean["Description"]
    .fillna(df_clean["StockCode"].map(description_lookup))
)

missing_after = df_clean["Description"].isna().sum()

print("Missing descriptions before:", missing_before)
print("Missing descriptions after:", missing_after)
print("Descriptions recovered:", missing_before - missing_after)

Missing descriptions before: 4275
Missing descriptions after: 363
Descriptions recovered: 3912


In [71]:
negative_price_count = (df_clean["Price"] < 0).sum()

df_clean = df_clean[df_clean["Price"] >= 0].copy()

print("Negative-price records removed:", negative_price_count)

Negative-price records removed: 5


In [72]:
df_clean["Customer ID"] = (
    df_clean["Customer ID"].astype("Int64")
)

print(df_clean["Customer ID"].dtype)

Int64


In [73]:
df_clean["InvoiceDate"] = pd.to_datetime(
    df_clean["InvoiceDate"]
)

print(df_clean["InvoiceDate"].dtype)

datetime64[us]


In [74]:
df_clean["IsCancellation"] = (
    df_clean["Invoice"]
    .astype(str)
    .str.startswith("C")
)

print(
    "Cancellation records:",
    df_clean["IsCancellation"].sum()
)

Cancellation records: 19104


In [75]:
df_clean["IsZeroPrice"] = (
    df_clean["Price"] == 0
)

print(
    "Zero-price records:",
    df_clean["IsZeroPrice"].sum()
)

Zero-price records: 6014


In [76]:
df_clean["TransactionType"] = "Sale"

In [77]:
df_clean.loc[
    df_clean["IsCancellation"],
    "TransactionType"
] = "Cancellation/Return"

In [78]:
df_clean["TransactionType"].value_counts()

TransactionType
Sale                   1013927
Cancellation/Return      19104
Name: count, dtype: int64

In [79]:
df_clean["Revenue"] = (
    df_clean["Quantity"] * df_clean["Price"]
)

In [80]:
df_clean[
    ["Quantity", "Price", "Revenue"]
].head(10)

,Quantity,Price,Revenue
0,12,6.95,83.4
1,12,6.75,81.0
2,12,6.75,81.0
3,48,2.10,100.8
4,24,1.25,30.0
5,24,1.65,39.6
6,24,1.25,30.0
7,10,5.95,59.5
8,12,2.55,30.6
9,12,3.75,45.0


In [81]:
print("Cleaned dataset shape:", df_clean.shape)

Cleaned dataset shape: (1033031, 12)


In [82]:
print("\nMissing values:")
print(df_clean.isnull().sum())


Missing values:
Invoice                 0
StockCode               0
Description           363
Quantity                0
InvoiceDate             0
Price                   0
Customer ID        235146
Country                 0
IsCancellation          0
IsZeroPrice             0
TransactionType         0
Revenue                 0
dtype: int64


In [83]:
print("\nData types:")
print(df_clean.dtypes)


Data types:
Invoice                    object
StockCode                  object
Description                object
Quantity                    int64
InvoiceDate        datetime64[us]
Price                     float64
Customer ID                 Int64
Country                       str
IsCancellation               bool
IsZeroPrice                  bool
TransactionType               str
Revenue                   float64
dtype: object


In [84]:
print("\nDuplicate rows:")
print(df_clean.duplicated().sum())


Duplicate rows:
0


In [85]:
cleaning_summary = pd.DataFrame({
    "Metric": [
        "Original Rows",
        "Rows After Cleaning",
        "Duplicate Rows Removed",
        "Negative Price Rows Removed",
        "Missing Descriptions Initially",
        "Descriptions Recovered",
        "Descriptions Still Missing",
        "Missing Customer IDs",
        "Zero Price Records Retained",
        "Negative Quantity Records Retained"
    ],
    "Value": [
        len(df_raw),
        len(df_clean),
        df_raw.duplicated().sum(),
        negative_price_count,
        4382,
        4019,
        df_clean["Description"].isna().sum(),
        df_clean["Customer ID"].isna().sum(),
        df_clean["IsZeroPrice"].sum(),
        (df_clean["Quantity"] < 0).sum()
    ]
})

cleaning_summary

,Metric,Value
0,Original Rows,1067371
1,Rows After Cleaning,1033031
2,Duplicate Rows Removed,34335
3,Negative Price Rows Removed,5
4,Missing Descriptions Initially,4382
5,Descriptions Recovered,4019
6,Descriptions Still Missing,363
7,Missing Customer IDs,235146
8,Zero Price Records Retained,6014
9,Negative Quantity Records Retained,22496


In [86]:
print("========== DATA VALIDATION ==========")

print(
    "Duplicate rows:",
    df_clean.duplicated().sum()
)

print(
    "Negative prices:",
    (df_clean["Price"] < 0).sum()
)

print(
    "Zero prices:",
    (df_clean["Price"] == 0).sum()
)

print(
    "Missing descriptions:",
    df_clean["Description"].isna().sum()
)

print(
    "Missing Customer IDs:",
    df_clean["Customer ID"].isna().sum()
)

print(
    "Negative quantities:",
    (df_clean["Quantity"] < 0).sum()
)

print(
    "Invalid dates:",
    df_clean["InvoiceDate"].isna().sum()
)

========== DATA VALIDATION ==========
Duplicate rows: 0
Negative prices: 0
Zero prices: 6014
Missing descriptions: 363
Missing Customer IDs: 235146
Negative quantities: 22496
Invalid dates: 0


In [87]:
output_path = "../data/cleaned/online_retail_II_cleaned.csv"

df_clean.to_csv(
    output_path,
    index=False
)

print("Cleaned dataset saved successfully!")
print(output_path)

Cleaned dataset saved successfully!
../data/cleaned/online_retail_II_cleaned.csv


In [88]:
print("Raw shape:", df_raw.shape)
print("Cleaned shape:", df_clean.shape)

print("\nDuplicates:", df_clean.duplicated().sum())
print("Negative prices:", (df_clean["Price"] < 0).sum())
print("Missing descriptions:", df_clean["Description"].isna().sum())
print("Missing Customer IDs:", df_clean["Customer ID"].isna().sum())
print("Zero prices:", (df_clean["Price"] == 0).sum())
print("Negative quantities:", (df_clean["Quantity"] < 0).sum())

print("\nTransaction types:")
print(df_clean["TransactionType"].value_counts())

Raw shape: (1067371, 8)
Cleaned shape: (1033031, 12)

Duplicates: 0
Negative prices: 0
Missing descriptions: 363
Missing Customer IDs: 235146
Zero prices: 6014
Negative quantities: 22496

Transaction types:
TransactionType
Sale                   1013927
Cancellation/Return      19104
Name: count, dtype: int64


In [89]:
rows_removed = len(df_raw) - len(df_clean)
retention_rate = len(df_clean) / len(df_raw) * 100

print("========== CLEANING SUMMARY ==========")
print(f"Original rows       : {len(df_raw):,}")
print(f"Cleaned rows        : {len(df_clean):,}")
print(f"Rows removed        : {rows_removed:,}")
print(f"Data retained       : {retention_rate:.2f}%")

========== CLEANING SUMMARY ==========
Original rows       : 1,067,371
Cleaned rows        : 1,033,031
Rows removed        : 34,340
Data retained       : 96.78%


In [90]:
print(
    df_clean.loc[
        (df_clean["Quantity"] < 0) &
        (~df_clean["IsCancellation"]),
        "TransactionType"
    ].value_counts()
)


TransactionType
Sale    3393
Name: count, dtype: int64


In [91]:
print(
    "Negative non-cancellation rows:",
    (
        (df_clean["Quantity"] < 0) &
        (~df_clean["IsCancellation"])
    ).sum()
)

Negative non-cancellation rows: 3393


In [92]:
df_clean.loc[
    (df_clean["Quantity"] < 0) &
    (~df_clean["IsCancellation"]),
    "TransactionType"
] = "Adjustment"

In [93]:
df_clean["TransactionType"].value_counts()

TransactionType
Sale                   1010534
Cancellation/Return      19104
Adjustment                3393
Name: count, dtype: int64

In [94]:
final_cleaning_summary = pd.DataFrame({
    "Metric": [
        "Original Records",
        "Final Records",
        "Records Removed",
        "Data Retention Rate",
        "Exact Duplicates Removed",
        "Negative Price Records Removed",
        "Missing Customer IDs",
        "Missing Descriptions",
        "Recovered Descriptions",
        "Zero Price Records",
        "Negative Quantity Records",
        "Cancellation/Return Records",
        "Adjustment Records",
        "Sale Records"
    ],
    "Value": [
        len(df_raw),
        len(df_clean),
        len(df_raw) - len(df_clean),
        f"{len(df_clean) / len(df_raw) * 100:.2f}%",
        df_raw.duplicated().sum(),
        negative_price_count,
        df_clean["Customer ID"].isna().sum(),
        df_clean["Description"].isna().sum(),
        4019,
        df_clean["IsZeroPrice"].sum(),
        (df_clean["Quantity"] < 0).sum(),
        (df_clean["TransactionType"] == "Cancellation/Return").sum(),
        (df_clean["TransactionType"] == "Adjustment").sum(),
        (df_clean["TransactionType"] == "Sale").sum()
    ]
})

final_cleaning_summary


,Metric,Value
0,Original Records,1067371
1,Final Records,1033031
2,Records Removed,34340
3,Data Retention Rate,96.78%
4,Exact Duplicates Removed,34335
5,Negative Price Records Removed,5
6,Missing Customer IDs,235146
7,Missing Descriptions,363
8,Recovered Descriptions,4019
9,Zero Price Records,6014


In [95]:
final_cleaning_summary.to_csv(
    "../output/cleaning_summary.csv",
    index=False
)

print("Cleaning summary exported successfully.")

Cleaning summary exported successfully.


In [96]:
print("========== FINAL VALIDATION ==========")

checks = {
    "Duplicate rows": df_clean.duplicated().sum(),
    "Negative prices": (df_clean["Price"] < 0).sum(),
    "Missing descriptions": df_clean["Description"].isna().sum(),
    "Missing Customer IDs": df_clean["Customer ID"].isna().sum(),
    "Zero prices": (df_clean["Price"] == 0).sum(),
    "Negative quantities": (df_clean["Quantity"] < 0).sum(),
    "Missing dates": df_clean["InvoiceDate"].isna().sum(),
    "Missing revenue": df_clean["Revenue"].isna().sum()
}

for check, value in checks.items():
    print(f"{check:<25}: {value}")

========== FINAL VALIDATION ==========
Duplicate rows           : 0
Negative prices          : 0
Missing descriptions     : 363
Missing Customer IDs     : 235146
Zero prices              : 6014
Negative quantities      : 22496
Missing dates            : 0
Missing revenue          : 0


In [97]:
df_clean["Year"] = df_clean["InvoiceDate"].dt.year
df_clean["Month"] = df_clean["InvoiceDate"].dt.month
df_clean["MonthName"] = df_clean["InvoiceDate"].dt.month_name()

In [98]:
df_clean["YearMonth"] = (
    df_clean["InvoiceDate"].dt.to_period("M").astype(str)
)

In [99]:
df_clean.to_csv(
    "../data/cleaned/online_retail_II_cleaned.csv",
    index=False
)

print("Final cleaned dataset saved.")

Final cleaned dataset saved.


In [100]:
df_clean["TransactionType"].value_counts()

TransactionType
Sale                   1010534
Cancellation/Return      19104
Adjustment                3393
Name: count, dtype: int64

In [101]:
final_cleaning_summary

,Metric,Value
0,Original Records,1067371
1,Final Records,1033031
2,Records Removed,34340
3,Data Retention Rate,96.78%
4,Exact Duplicates Removed,34335
5,Negative Price Records Removed,5
6,Missing Customer IDs,235146
7,Missing Descriptions,363
8,Recovered Descriptions,4019
9,Zero Price Records,6014


# Final Summary

## Dataset Before Cleaning

- Records: 1,067,371
- Columns: 8

## Data Quality Issues Identified

- 34,335 exact duplicate records
- 243,007 missing Customer IDs
- 4,382 missing product descriptions
- 22,950 negative-quantity records
- 6,202 zero-price records
- 5 negative-price accounting adjustment records

## Cleaning Actions

- Removed 34,335 exact duplicate records.
- Removed 5 negative-price "Adjust bad debt" records.
- Recovered 4,019 missing descriptions using StockCode-based mapping.
- Retained 363 descriptions that could not be safely recovered.
- Retained missing Customer IDs because the associated transactions remain useful for analysis.
- Converted Customer ID from float64 to nullable Int64.
- Preserved negative quantities and classified transactions into Sale, Cancellation/Return, and Adjustment.
- Added IsCancellation and IsZeroPrice flags.
- Added Revenue as Quantity × Price.
- Prepared date-related fields for subsequent analysis.

## Final Dataset

- Final records: 1,033,031
- Records retained: 96.78%
- Duplicate records: 0
- Negative prices: 0

In [102]:
print(df_clean.columns.tolist())

['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country', 'IsCancellation', 'IsZeroPrice', 'TransactionType', 'Revenue', 'Year', 'Month', 'MonthName', 'YearMonth']


In [104]:
df_clean.to_csv(
    "../data/cleaned/online_retail_II_cleaned.csv",
    index=False
)

print("Final cleaned dataset saved successfully.")

Final cleaned dataset saved successfully.


In [105]:
print(df_clean.columns.tolist())

['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country', 'IsCancellation', 'IsZeroPrice', 'TransactionType', 'Revenue', 'Year', 'Month', 'MonthName', 'YearMonth']


In [106]:
print(df_clean.dtypes)

Invoice                    object
StockCode                  object
Description                object
Quantity                    int64
InvoiceDate        datetime64[us]
Price                     float64
Customer ID                 Int64
Country                       str
IsCancellation               bool
IsZeroPrice                  bool
TransactionType               str
Revenue                   float64
Year                        int32
Month                       int32
MonthName                     str
YearMonth                     str
dtype: object
